### ISSUE TITLE
[Research] Contextual Distillation: Internalizing Document Knowledge via Teacher-Student Distillation on Generated QA Pairs

### ISSUE THREAD

**Issue by LiamConnell** [karpathy/nanochat#471]
## Summary

A training methodology to "internalize" document/context knowledge directly into model weights, eliminating the need for context retrieval at inference time. The approach uses teacher-student distillation on synthetically generated question-answer pairs derived from the source document.

## Motivation

Current approaches to incorporating external knowledge into LLMs typically require:
- **RAG** (Retrieval-Augmented Generation) - adds latency and retrieval complexity
- **Fine-tuning on raw documents** - often leads to poor knowledge retention and catastrophic forgetting
- **Continued pretraining** - expensive and doesn't guarantee knowledge accessibility

Contextual distillation offers a middle ground: targeted knowledge injection that teaches the model *how to answer* about specific content, not just exposing it to the content.

### Phase 1: Question Generation
Given a context document `C`, generate `N` diverse questions `{Q_1, Q_2, ..., Q_N}` that:
- Require knowledge from `C` to answer correctly
- Cover different aspects/sections of the document
- Vary in complexity (factual recall, reasoning, synthesis)

```
Input: Context C
Output: Questions Q_1...Q_N where answering Q_i requires C
```

### Phase 2: Teacher Answer Generation
Using a teacher model with full access to context `C`:
- Generate answers `A_i` for each question `Q_i`
- Record token-level log-probabilities `logP_teacher(a_t | Q_i, C, a_{<t})` for each answer

```
Teacher Input: [C, Q_i]
Teacher Output: A_i with per-token log-probs
```

### Phase 3: Student Distillation
Train a student model **without** access to `C`:
- Input: Only the question `Q_i`
- Target: Match the teacher's output distribution (via log-prob alignment)

**Loss function options:**
1. **KL Divergence**: Minimize `KL(P_teacher || P_student)` at token level
2. **Forward KL**: `sum_t -P_teacher(a_t) * log(P_student(a_t))`
3. **Soft label distillation**: Use teacher logits as soft targets with temperature

```python
# Pseudocode
for Q_i, A_i, teacher_logprobs in dataset:
    student_logits = student_model(Q_i)  # No context!
    loss = distillation_loss(student_logits, teacher_logprobs, A_i)
    loss.backward()
```

## Key Design Decisions

| Decision | Options | Notes |
|----------|---------|-------|
| Question generation | Same model / Separate model | Could use teacher or specialized QA-gen model |
| Number of questions N | Fixed / Adaptive | Scale with document length/complexity |
| Question diversity | Random / Stratified / Curriculum | Ensure coverage of document |
| Distillation target | Hard labels / Soft labels / Full distribution | Trade-off: signal richness vs. compute |
| Student initialization | Base model / Teacher copy | Affects training dynamics |

## Expected Benefits

1. **Inference efficiency**: No retrieval needed, knowledge is in weights
2. **Targeted injection**: QA format ensures knowledge is *accessible*, not just stored
3. **Controllable**: Choose exactly what knowledge to internalize
4. **Composable**: Can distill multiple documents sequentially or in parallel

## Potential Challenges

- **Capacity limits**: How much context can be internalized before interference?
- **Forgetting**: Does internalizing new context degrade existing capabilities?
- **Question quality**: Generated questions may miss important knowledge
- **Generalization**: Will the model answer paraphrased questions correctly?

## Experiments to Run

1. **Baseline comparison**: RAG vs. fine-tuning vs. contextual distillation on same docs
2. **Scaling**: Vary N (questions per doc) and measure knowledge retention
3. **Evaluation**: Held-out questions about the same document
4. **Forgetting analysis**: Track performance on unrelated benchmarks during training
5. **Ablations**: Question diversity strategies, distillation loss functions

## Implementation Steps

- [ ] Question generation pipeline (given doc → diverse QA pairs)
- [ ] Teacher inference with log-prob extraction
- [ ] Distillation training loop
- [ ] Evaluation framework (in-domain QA accuracy, out-of-domain retention)
- [ ] Experiment tracking and analysis

## References

- Knowledge Distillation (Hinton et al., 2015)
- Self-Instruct / Synthetic data generation
- Document-grounded QA generation

**Comment by LiamConnell** [comment]
Closing - created on wrong repository by mistake.
